commit c154ce33074bb1952a43d285b84c1cd6e9a5cf26
Author: Vitaliy Filippov <vitalif@yourcmc.ru>
Date:   Fri Nov 16 02:59:19 2012 +0400

    Patch: mergeconflicts-extension
    Type: feature
    
    Patch needed for MergeConflicts extension. Allows to change merge
    conflict display style. CustIS Bug 80143.

diff --git a/includes/DefaultSettings.php b/includes/DefaultSettings.php
index 7518c77..d83aa01 100644
--- a/includes/DefaultSettings.php
+++ b/includes/DefaultSettings.php
@@ -6390,6 +6390,9 @@ $wgSeleniumConfigFile = null;
 $wgDBtestuser = ''; //db user that has permission to create and drop the test databases only
 $wgDBtestpassword = '';
 
+/** Defines that MergeConflicts extension patch was applied to this MW installation */
+define ( 'MW_PATCH_MERGE_CONFLICTS', 1 );
+
 /**
  * Associative array mapping namespace IDs to the name of the content model pages in that namespace should have by
  * default (use the CONTENT_MODEL_XXX constants). If no special content type is defined for a given namespace,
diff --git a/includes/EditPage.php b/includes/EditPage.php
index 8b2dbb5..d5db54f 100644
--- a/includes/EditPage.php
+++ b/includes/EditPage.php
@@ -1592,12 +1592,15 @@ class EditPage {
 				$content = $textbox_content; // do not try to merge here!
 			} elseif ( $this->isConflict ) {
 				# Attempt merge
-				if ( $this->mergeChangesIntoContent( $content ) ) {
+				if ( $res = $this->mergeChangesIntoContent( $content, true ) ) {
 					// Successful merge! Maybe we should tell the user the good news?
 					$this->isConflict = false;
 					wfDebug( __METHOD__ . ": Suppressing edit conflict, successful merge.\n" );
 				} else {
 					$this->section = '';
+					// was merging available?
+					$this->mMergeAvailable = $res !== NULL;
+					$this->textbox2 = $this->textbox1;
 					$this->textbox1 = ContentHandler::getContentText( $content );
 					wfDebug( __METHOD__ . ": Keeping edit conflict, failed merge.\n" );
 				}
@@ -1788,7 +1791,7 @@ class EditPage {
 	 *
 	 * @return bool
 	 */
-	private function mergeChangesIntoContent( &$editContent ) {
+	private function mergeChangesIntoContent( &$editContent, $force = false ) {
 		wfProfileIn( __METHOD__ );
 
 		$db = wfGetDB( DB_MASTER );
@@ -1813,7 +1816,12 @@ class EditPage {
 
 		$handler = ContentHandler::getForModelID( $baseContent->getModel() );
 
-		$result = $handler->merge3( $baseContent, $editContent, $currentContent );
+		$result = $handler->merge3( $baseContent, $editContent, $currentContent, $force );
+
+		if ( $force ) {
+			$editContent = $result[0];
+			return $result[1];
+		}
 
 		if ( $result ) {
 			$editContent = $result;
@@ -2207,15 +2215,6 @@ class EditPage {
 		}
 
 		if ( $this->isConflict ) {
-			// In an edit conflict bypass the overridable content form method
-			// and fallback to the raw wpTextbox1 since editconflicts can't be
-			// resolved between page source edits and custom ui edits using the
-			// custom edit ui.
-			$this->textbox2 = $this->textbox1;
-
-			$content = $this->getCurrentContent();
-			$this->textbox1 = $this->toEditText( $content );
-
 			$this->showTextbox1();
 		} else {
 			$this->showContentForm();
@@ -2285,7 +2284,8 @@ class EditPage {
 		$wgOut->addHTML( implode( "\n", $this->mTitle->getEditNotices() ) );
 
 		if ( $this->isConflict ) {
-			$wgOut->wrapWikiMsg( "<div class='mw-explainconflict'>\n$1\n</div>", 'explainconflict' );
+			$wgOut->wrapWikiMsg( "<div class='mw-explainconflict'>\n$1\n</div>",
+				$this->mMergeAvailable ? 'explainconflictmerged' : 'explainconflict' );
 			$this->edittime = $this->mArticle->getTimestamp();
 		} else {
 			if ( $this->section != '' && !$this->isSectionEditSupported() ) {
@@ -2608,11 +2608,11 @@ HTML
 		$this->showTextbox( $textoverride !== null ? $textoverride : $this->textbox1, 'wpTextbox1', $attribs );
 	}
 
-	protected function showTextbox2() {
+	public function showTextbox2() {
 		$this->showTextbox( $this->textbox2, 'wpTextbox2', array( 'tabindex' => 6, 'readonly' ) );
 	}
 
-	protected function showTextbox( $text, $name, $customAttribs = array() ) {
+	public function showTextbox( $text, $name, $customAttribs = array() ) {
 		global $wgOut, $wgUser;
 
 		$wikitext = $this->safeUnicodeOutput( $text );
@@ -2859,6 +2859,7 @@ HTML
 		global $wgOut;
 
 		if ( wfRunHooks( 'EditPageBeforeConflictDiff', array( &$this, &$wgOut ) ) ) {
+			$this->textbox2 = $this->getContent();
 			$wgOut->wrapWikiMsg( '<h2>$1</h2>', "yourdiff" );
 
 			$content1 = $this->toEditContent( $this->textbox1 );
@@ -2866,13 +2867,13 @@ HTML
 
 			$handler = ContentHandler::getForModelID( $this->contentModel );
 			$de = $handler->createDifferenceEngine( $this->mArticle->getContext() );
-			$de->setContent( $content2, $content1 );
+			$de->setContent( $content1, $content2 );
 			$de->showDiff(
 				wfMessage( 'yourtext' )->parse(),
 				wfMessage( 'storedversion' )->text()
 			);
 
-			$wgOut->wrapWikiMsg( '<h2>$1</h2>', "yourtext" );
+			$wgOut->wrapWikiMsg( '<h2>$1</h2>', "storedversion" );
 			$this->showTextbox2();
 		}
 	}
diff --git a/includes/GlobalFunctions.php b/includes/GlobalFunctions.php
index 016736f..d76d96d 100644
--- a/includes/GlobalFunctions.php
+++ b/includes/GlobalFunctions.php
@@ -2932,7 +2932,7 @@ function wfMerge( $old, $mine, $yours, &$result ) {
 
 	if( !$haveDiff3 ) {
 		wfDebug( "diff3 not found\n" );
-		return false;
+		return NULL;
 	}
 
 	# Make temporary files
@@ -2941,6 +2941,10 @@ function wfMerge( $old, $mine, $yours, &$result ) {
 	$mytextFile = fopen( $mytextName = tempnam( $td, 'merge-mine-' ), 'w' );
 	$yourtextFile = fopen( $yourtextName = tempnam( $td, 'merge-your-' ), 'w' );
 
+	if ($old{strlen($old)-1} != "\n") $old .= "\n";
+	if ($mine{strlen($mine)-1} != "\n") $mine .= "\n";
+	if ($yours{strlen($yours)-1} != "\n") $yours .= "\n";
+
 	# NOTE: diff3 issues a warning to stderr if any of the files does not end with
 	#       a newline character. To avoid this, we normalize the trailing whitespace before
 	#       creating the diff.
@@ -2952,23 +2956,11 @@ function wfMerge( $old, $mine, $yours, &$result ) {
 	fwrite( $yourtextFile, rtrim( $yours ) . "\n" );
 	fclose( $yourtextFile );
 
-	# Check for a conflict
-	$cmd = wfEscapeShellArg( $wgDiff3 ) . ' -a --overlap-only ' .
-		wfEscapeShellArg( $mytextName ) . ' ' .
-		wfEscapeShellArg( $oldtextName ) . ' ' .
-		wfEscapeShellArg( $yourtextName );
-	$handle = popen( $cmd, 'r' );
-
-	if( fgets( $handle, 1024 ) ) {
-		$conflict = true;
-	} else {
-		$conflict = false;
-	}
-	pclose( $handle );
-
 	# Merge differences
-	$cmd = wfEscapeShellArg( $wgDiff3 ) . ' -a -e --merge ' .
-		wfEscapeShellArg( $mytextName, $oldtextName, $yourtextName );
+	$cmd = wfEscapeShellArg( $wgDiff3 ) . ' -a -A --merge ' . wfEscapeShellArg(
+		'-L', wfMsg( 'merge-mine' ), '-L', wfMsg( 'merge-old' ),
+		'-L', wfMsg( 'merge-their' ), $mytextName, $oldtextName, $yourtextName
+	);
 	$handle = popen( $cmd, 'r' );
 	$result = '';
 	do {
@@ -2978,7 +2970,7 @@ function wfMerge( $old, $mine, $yours, &$result ) {
 		}
 		$result .= $data;
 	} while ( true );
-	pclose( $handle );
+	$conflict = pclose( $handle ) != 0;
 	unlink( $mytextName );
 	unlink( $oldtextName );
 	unlink( $yourtextName );
diff --git a/includes/content/ContentHandler.php b/includes/content/ContentHandler.php
index 9c20195..e406baf 100644
--- a/includes/content/ContentHandler.php
+++ b/includes/content/ContentHandler.php
@@ -701,7 +701,7 @@ abstract class ContentHandler {
 	 *
 	 * @return Content|Bool
 	 */
-	public function merge3( Content $oldContent, Content $myContent, Content $yourContent ) {
+	public function merge3( Content $oldContent, Content $myContent, Content $yourContent, $force = false ) {
 		return false;
 	}
 
diff --git a/includes/content/TextContentHandler.php b/includes/content/TextContentHandler.php
index e7f41e1..512c01e 100644
--- a/includes/content/TextContentHandler.php
+++ b/includes/content/TextContentHandler.php
@@ -61,7 +61,7 @@ class TextContentHandler extends ContentHandler {
 	 *
 	 * @return Content|Bool
 	 */
-	public function merge3( Content $oldContent, Content $myContent, Content $yourContent ) {
+	public function merge3( Content $oldContent, Content $myContent, Content $yourContent, $force = false ) {
 		$this->checkModelID( $oldContent->getModel() );
 		$this->checkModelID( $myContent->getModel() );
 		$this->checkModelID( $yourContent->getModel() );
@@ -74,7 +74,7 @@ class TextContentHandler extends ContentHandler {
 
 		$ok = wfMerge( $old, $mine, $yours, $result );
 
-		if ( !$ok ) {
+		if ( !$ok && !$force ) {
 			return false;
 		}
 
@@ -83,6 +83,9 @@ class TextContentHandler extends ContentHandler {
 		}
 
 		$mergedContent = $this->unserializeContent( $result, $format );
+		if ( $force ) {
+			return array( $mergedContent, $ok );
+		}
 		return $mergedContent;
 	}
 
